lib/clAdapters/google/errors.ts

```diff
@@ -0,0 +1,46 @@
+// Type union instead of enum so calling code can get name without type details
+export type GoogleErrorType =
+  'InvalidGrant' |
+  'UnauthorizedClient'
+;
+
+export interface InvalidGrant {
+  kind: 'InvalidGrant';
+  err: Error;
+}
+
+export interface UnauthorizedClient {
+  kind: 'UnauthorizedClient';
+  err: Error;
+}
+
+export type GoogleError = InvalidGrant | UnauthorizedClient;
+
+export function createGoogleError<T extends GoogleErrorType>(
+  kind: T,
+  err?: Error
+): GoogleError {
+  if (!err) {
+    err = new Error();
+    Error.captureStackTrace(err, createGoogleError);
+  }
+
+  // return {
+  //     kind,
+  //     err
+  // };
+  // Type value is not enforceable:
```

Comment:
This is because you've declared an explicit return type above, if you instead leave the return type to be inferred, typescript should do the right thing. [See this playground example I made](http://www.typescriptlang.org/play/index.html#src=type%20GoogleErrorType%0D%0A%20%20%20%20%3D%20'invalid'%0D%0A%20%20%20%20%7C%20'unauthorized'%3B%0D%0A%0D%0Atype%20GoogleError%0D%0A%20%20%3D%20InvalidGrant%0D%0A%20%20%7C%20UnauthorizedClient%3B%0D%0A%0D%0A%0D%0Ainterface%20InvalidGrant%20%7B%0D%0A%20%20%20%20kind%3A%20'invalid'%3B%0D%0A%20%20%20%20err%3A%20Error%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20UnauthorizedClient%20%7B%0D%0A%20%20%20%20kind%3A%20'unauthorized'%3B%0D%0A%20%20%20%20err%3A%20Error%3B%0D%0A%7D%0D%0A%0D%0A%0D%0A%0D%0A%2F**%0D%0A%20*%0D%0A%20*%0D%0A%20*%20example%20of%20correct%20return%20types%0D%0A%20*%0D%0A%20*%0D%0A%20*%2F%0D%0A%0D%0A%2F**%0D%0A%20*%20this%20is%20%22structurally%22%20an%20invalid%20grant%2C%20so%20we%20can%0D%0A%20*%20use%20contextual%20typing%20to%20assert%0D%0A%20*%2F%0D%0Aconst%20invalid%3A%20InvalidGrant%20%3D%20createError('invalid')%3B%0D%0A%0D%0A%2F**%0D%0A%20*%20note%2C%20this%20fails%20to%20compile%2C%20meaning%20our%20context%20based%20on%20the%20provided%20type%0D%0A%20*%20to%20createError%20results%20in%20the%20specific%20return%20type%0D%0A%20*%2F%0D%0Aconst%20unauthed%3A%20UnauthorizedClient%20%3D%20createError('invalid')%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%2F**%0D%0A%20*%20%20creating%20errors%0D%0A%20*%2F%0D%0A%2F**%0D%0A%20*%20%20if%20we%20want%20to%20have%20return%20types%20explicitly%20return%0D%0A%20*%20%20the%20correct%20interface%20based%20on%20the%20kind%0D%0A%20*%20%20(instead%20of%20just%20structurally%20correct%20types)%2C%0D%0A%20*%20%20we%20can%20add%20overloads%20like%20this%0D%0A%20*%2F%0D%0A%2F%2F%20uncomment%20below%20to%20switch%20the%20call%20to%20createError%20to%20explicitly%20return%20InvalidGrant%0D%0A%2F%2F%20function%20createError(kind%3A%20'invalid'%2C%20err%3F%3A%20Error)%3A%20InvalidGrant%3B%0D%0Afunction%20createError%3CT%20extends%20GoogleErrorType%3E(%0D%0A%20%20kind%3A%20T%2C%0D%0A%20%20err%3A%20Error%20%3D%20new%20Error()%0D%0A)%20%7B%0D%0A%20%20return%20%7B%20kind%2C%20err%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%2F**%0D%0A%20*%20example%20desriminating%0D%0A%20*%2F%0D%0Afunction%20handleGoogleError(googErr%3A%20GoogleError)%20%7B%0D%0A%20%20%20%20switch%20(googErr.kind)%20%7B%0D%0A%20%20%20%20%20%20%20%20case%20GoogleErrorType.InvalidGrant%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F**%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%20here%20the%20type%20system%20knows%20that%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%20the%20error%20is%20an%20invalid%20grant%2C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%20as%20this%20compiles%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%2F%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20let%20e%3A%20InvalidGrant%20%3D%20googErr%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D)